compiler/src/dotty/tools/dotc/transform/LazyVals.scala

```diff
@@ -73,7 +73,18 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
     else {
       val isField = sym.owner.isClass
       if (isField)
-        if (sym.isAllOf(SyntheticModule))
+        if sym.isAllOf(SyntheticModule)
+           && sym.allOverriddenSymbols.isEmpty
+           && !sym.name.is(ExpandedName) then
+          // I am not sure what the conditions for this optimization should be.
+          // It was applied for all synthetic objects, but this is clearly false, as t704 demonstrates.
+          // It seems we have to at least exclude synthetic objects that derive from mixins.
+          // This is done by demanding that the object does not override anything.
+          // Figuring out whether a symbol implements a trait module is not so simple.
+          // For non-private trait members we can check whether there is an overridden symbol.
+          // For private trait members this does not work, since `ensureNotPrivate` in phase Mixins
+          // does change the name but does not update the owner's scope, so `allOverriddenSymbols` does
+          // not work in that case. However, we can check whether the name is an ExpandedName instead.
```

Comment:
I think the code is a little brittle.

The optimization was introduced in https://github.com/lampepfl/dotty/commit/8f47f1694a283c0dc280913eb0c6a895ebd7ef1e to fix #676 . Since #4090, we no longer generate companion objects in FirstTransform anymore, it seems we can just disable the optimization.

If the optimization is needed, what about using a compiler flag to flag an object as _eager_, and we only check flags here? That flag could also be reused to enable more optimization opportunities without worrying about semantic problems.

Reply:
Even with automatic generation of companion objects disabled, I believe there are still a lot of them, so it's worthwhile to optimize. But the criterion as it stands is somewhat arbitrary and could be improved. E.g. we could also strictify companion objects that are pure.